Add support for target annotations and labels

terricain opened the pull request:
This adds template fields to specify labels and annotations for the target configmaps and secrets along with some extra validation logic to prevent clobbering `trust.cert-manager.io` labels and annotations.

```yaml
apiVersion: trust.cert-manager.io/v1alpha1
kind: Bundle
metadata:
  name: test
spec:
  sources:
    ...
  target:
    secretTemplate:
      labels: {}
      annotations: {}
    configMapTemplate:
      labels: {}
      annotations: {}
```

@erikgb before I go and write some tests, do we like these field names? or would the following look better? Naming things is the worst, if someone has any better ideas then I'm all for it 😆 

```yaml
apiVersion: trust.cert-manager.io/v1alpha1
kind: Bundle
metadata:
  name: test
spec:
  sources:
    ...
  target:
    configMap:
      key: "root-certs.pem"
      template:
        labels: {}
        annotations: {}
    secret:
      key: "root-certs.pem"
      template:
        labels: {}
        annotations: {}
```

---

Fixes #23

erikgb commented:
Yeah, naming is the hardest thing we do. 🤔 And this is mixed with API design.

I am working on the next version of the bundle API: https://github.com/cert-manager/trust-manager/pull/486. It might be worth checking that this fits in there also.

Another slightly modified suggestion from me:

`````yaml
apiVersion: trust.cert-manager.io/v1alpha1
kind: Bundle
metadata:
  name: test
spec:
  sources:
    ...
  target:
    configMap:
      key: "root-certs.pem"
      metadata:
        labels: {}
        annotations: {}
    secret:
      key: "root-certs.pem"
      metadata:
        labels: {}
        annotations: {}
````

I'll bring this up on today's stand-up. Would you like to join? It's at 10:30 UK time. 😺

erikgb commented:
@terricain, we discussed this PR in today's stand-up, and we think we prefer my suggestion in https://github.com/cert-manager/trust-manager/pull/582#issuecomment-2785322812. Do you want to become a member of our organization? This will, among others, allow your PRs to be enabled for CI without approval from other members. How to become a member is described here: https://github.com/cert-manager/community/blob/main/GOVERNANCE.md#becoming-a-github-member. I would be happy to sponsor your membership!

terricain commented:
I'd have joined the standup if I had read this in time 😄  (also how does one join the standup? I've only found links to a biweekly zoom meeting)

Yep I'll raise an issue to become an org member once I find a 2nd to sponsor it, I think you've got to all my PR's so far so haven't interacted with anyone else yet.

https://github.com/cert-manager/trust-manager/pull/582#issuecomment-2785322812 agreed, this looks much better than what I initially came up with, and should transfer nicely to the proposed API version in #486 (+1 for renaming it to ClusterBundle, makes the fact its not a namespaced resource much more obvious) 

Is it worth fixing up this PR and getting it merged or waiting till the new API is merged?

erikgb commented:
> I'd have joined the standup if I had read this in time 😄 (also how does one join the standup? I've only found links to a biweekly zoom meeting)

I have done this only once, but I think you will get invites to all meetings by joining our mailing lists, ref. https://cert-manager.io/docs/contributing/#meetings.

> Yep I'll raise an issue to become an org member once I find a 2nd to sponsor it, I think you've got to all my PR's so far so haven't interacted with anyone else yet.

I am pretty sure we should be able to find another sponsor. Didn't you talk with @maelvls at KubeCon ContribFest?

> Is it worth fixing up this PR and getting it merged or waiting till the new API is merged?

Please finish this PR! The new API is still far from finalized, and this PR will fix a requested feature.

terricain commented:
Updated API to use:
```
apiVersion: trust.cert-manager.io/v1alpha1
kind: Bundle
metadata:
  name: test
spec:
  sources:
    ...
  target:
    configMap:
      key: "root-certs.pem"
      metadata:
        labels: {}
        annotations: {}
    secret:
      key: "root-certs.pem"
      metadata:
        labels: {}
        annotations: {}
```

Tried adding some CEL expressions to validate that the annotations and labels didn't start with prefixes we care about but couldn't seem to get it to work without it looking ugly so have left the regular webhook validations in.

The `KeySelector` struct was used for the JKS and PKCS12 parts of the API as well as the target ConfigMap/Secret, it didn't make sense for the JKS/PKCS12 fields to have a `metadata` field so moved them to use a different struct with just `Key`. That and there were less occurrences of that to change than if I changed `KeySelector` for the ConfigMap/Secret. If there is a nicer way to achieve this let me know 😄.

The changes might be easier to look at commit-by-commit

---

Was getting some inconsistent connection refused errors with the ciphers test so added in a small loop there to wait for the webhook server to start.

erikgb commented:
/ok-to-test

erikgb commented:
> Was getting some inconsistent connection refused errors with the ciphers test so added in a small loop there to wait for the webhook server to start.

Yeah, I was also struggling with those errors yesterday. I will see if I can reproduce it now and create an issue for it. It should probably be fixed in a separate PR.

I will also try to review your PR tomorrow! Thanks!

terricain commented:
Thanks. Sure happy to pull it out and spin up another PR if you can reproduce it

erikgb commented on pkg/apis/trust/v1alpha1/types_bundle.go:
```diff
@@ -217,6 +217,29 @@ type KeySelector struct {
+}
+
+// KeySelectorWithoutMetadata is a reference to a key for some map data object without optional metadata
+type KeySelectorWithoutMetadata struct {
```
Even if your suggestion would change less code, I think we should avoid changing unrelated code (additional formats). So I created an alternative solution here: https://github.com/cert-manager/trust-manager/compare/main...erikgb:trust-manager:pr-582-suggestion. Please let me know what you think!

terricain replied:
Ah nice, yeah this looks good, I'll pull that in

terricain replied:
Done. I've added some getters on TargetTemplate which handle both if Metadata or annotations/labels are nil, is the `types_bundle.go` the best place, or is it worth splitting them out into another file?

erikgb commented on pkg/apis/trust/v1alpha1/types_bundle.go:
```diff
@@ -217,6 +217,29 @@ type KeySelector struct {
+
+	// Metadata is an optional set of labels and annotations to be copied to the target.
+	// +optional
+	Metadata *TargetMetadata `json:"metadata"`
```
```suggestion
	Metadata *TargetMetadata `json:"metadata,omitempty"`
```

erikgb commented on pkg/bundle/internal/target/target.go:
```diff
@@ -128,20 +128,22 @@ func (r *Reconciler) syncConfigMap(
-		WithAnnotations(map[string]string{
-			trustapi.BundleHashAnnotationKey: bundleHash,
-		}).
+		WithAnnotations(annotations).
+		WithLabels(labels).
```
```suggestion
		WithAnnotations(annotations).
		WithAnnotations(map[string]string{
			trustapi.BundleHashAnnotationKey: bundleHash,
                }).
		WithLabels(labels).
```

erikgb commented on pkg/bundle/internal/target/target.go:
```diff
@@ -128,20 +128,22 @@ func (r *Reconciler) syncConfigMap(
 	if !apierrors.IsNotFound(err) {
 		// Exit early if no update is needed
-		if exit, err := r.needsUpdate(ctx, target.Kind, targetObj, bundle, bundleHash); err != nil {
+		if exit, err := r.needsUpdate(ctx, target.Kind, targetObj, bundle, bundleHash, annotations, labels); err != nil {
```
```suggestion
		if exit, err := r.needsUpdate(ctx, target.Kind, targetObj, bundle, bundleHash, bundleTarget.ConfigMap.Metadata); err != nil {
```

terricain replied:
As you mentioned later on, ended up not needing this as WithAnnotations(annotations) handles this in conjunction with the added logic to the hash

erikgb commented on pkg/bundle/internal/target/target.go:
```diff
@@ -128,20 +128,22 @@ func (r *Reconciler) syncConfigMap(
 	}
 	binData := resolvedBundle.BinaryData
 
+	annotations, labels := mapsFromBundle(bundleTarget.ConfigMap.Metadata)
+	annotations[trustapi.BundleHashAnnotationKey] = bundleHash
+
```
```suggestion
```

erikgb commented on pkg/bundle/internal/target/target.go:
```diff
@@ -254,6 +258,34 @@ func (r *Reconciler) needsUpdate(ctx context.Context, kind Kind, obj *metav1.Par
+		}
+	}
+
+	// Find keys present in the target that should be removed
+	for annotationKey := range obj.GetAnnotations() {
+		if annotationKey == trustapi.BundleHashAnnotationKey {
+			continue
+		} else if _, exists := annotations[annotationKey]; !exists {
+			needsUpdate = true
+		}
+	}
+	for labelKey := range obj.GetLabels() {
+		if labelKey == trustapi.BundleLabelKey {
+			continue
+		} else if _, exists := labels[labelKey]; !exists {
+			needsUpdate = true
+		}
+	}
+
```
```suggestion
```

Since we are using server-side apply (SSA) in trust-manager, checking for labels/annotations to remove should not be required. But since we have this "needsUpdate" logic, and don't patch targets unconditionally, we need to create "smarter" logic. Probably based on the managed fields.

Other operators may add labels/annotations to targets, so we cannot just assume trust-manager is managing ALL labels/annotations.

Maybe @inteon has a good solution to this?

terricain replied:
Yep, updated the hash function to take target annotations/labels key & values into consideration if they exist, then the hash should always be different for any combination but also ignore any labels/annotations added by third parties.

erikgb commented on pkg/bundle/internal/target/target.go:
```diff
@@ -455,3 +487,17 @@ func TrustBundleHash(data []byte, additionalFormats *trustapi.AdditionalFormats)
 
 	return hex.EncodeToString(hashValue[:])
 }
+
+// mapsFromBundle returns the annotations and labels maps from a template ensuring the maps are not nil.
+func mapsFromBundle(targetMetadata *trustapi.TargetMetadata) (map[string]string, map[string]string) {
+	annotations := map[string]string{}
+	labels := map[string]string{}
+
+	if targetMetadata != nil && targetMetadata.Annotations != nil {
+		annotations = targetMetadata.Annotations
+	}
+	if targetMetadata != nil && targetMetadata.Labels != nil {
+		labels = targetMetadata.Labels
+	}
+	return annotations, labels
+}
```
```suggestion
```

Instead of this helper function, I would suggest creating nil-safe getters for `Labels` and `Annotations` on the `TargetMetadata` struct.

erikgb reviewed (commented):
Thanks for working on this @terricain! ❤️ I did an initial review pass and added a few comments/suggestions.

The most delicate part remaining is probably the "needsUpdate" logic. trust-manager is using server-side apply (SSA), and I have suggested unconditional patching of resources many times to avoid this complex and error-prone logic. But it's still there, so we need to find a good way of determining if target labels/annotations are up-to-date or not.

terricain commented:
Thanks for the review @erikgb, tons of good suggestions 🙂.

I updated the v1alpha2 ClusterBundle with what I added to v1alpha1 Bundle as it _looks_ like for now its just a copy that isn't really uesd?

I think I've addressed all the comments, I resolved a few that were outdated/easy fixes.

erikgb commented:
> I updated the v1alpha2 ClusterBundle with what I added to v1alpha1 Bundle as it _looks_ like for now its just a copy that isn't really used?

Yes, that's correct. `ClusterBundle` is now just a copy of `Bundle`, and the API generation is disabled. I have https://github.com/cert-manager/trust-manager/pull/486 in progress, but I will take this new feature into consideration when merged.
 
> I think I've addressed all the comments, I resolved a few that were outdated/easy fixes.

Cool, thanks! I will make another review pass soon.

erikgb commented on pkg/bundle/internal/target/target.go:
```diff
@@ -122,7 +122,7 @@ func (r *Reconciler) syncConfigMap(
 	// Generated PKCS #12 is not deterministic - best we can do here is update if the pem cert has
 	// changed (hence not checking if PKCS #12 matches)
-	bundleHash := TrustBundleHash([]byte(resolvedBundle.Data), bundle.Spec.Target.AdditionalFormats)
+	bundleHash := TrustBundleHash([]byte(resolvedBundle.Data), bundle.Spec.Target.AdditionalFormats, bundleTarget.ConfigMap)
```
```suggestion
	bundleHash := TrustBundleHash([]byte(resolvedBundle.Data), bundleTarget.AdditionalFormats, bundleTarget.ConfigMap)
```

Nit: A bit shorter.

erikgb commented on pkg/bundle/internal/target/target.go:
```diff
@@ -199,7 +201,7 @@ func (r *Reconciler) syncSecret(
 	// Generated PKCS #12 is not deterministic - best we can do here is update if the pem cert has
 	// changed (hence not checking if PKCS #12 matches)
-	bundleHash := TrustBundleHash([]byte(resolvedBundle.Data), bundle.Spec.Target.AdditionalFormats)
+	bundleHash := TrustBundleHash([]byte(resolvedBundle.Data), bundle.Spec.Target.AdditionalFormats, bundleTarget.Secret)
```
```suggestion
	bundleHash := TrustBundleHash([]byte(resolvedBundle.Data), bundleTarget.AdditionalFormats, bundleTarget.Secret)
```

Nit: A bit shorter.

erikgb commented on pkg/bundle/internal/target/target.go:
```diff
@@ -442,6 +446,16 @@ func TrustBundleHash(data []byte, additionalFormats *trustapi.AdditionalFormats)
 	}
 
+	// Add Target annotations and labels to the hash so it becomes aware of changes and triggers an update.
+	if target != nil {
```
I don't think `target` should be nil in this context. But maybe make `GetAnnotations` and `GetLabels` nil-safe, to make this a bit more compact?

erikgb commented on pkg/webhook/validation.go:
```diff
@@ -231,9 +233,49 @@ func (v *validator) validate(ctx context.Context, obj runtime.Object) (admission
+
+// validateAnnotationsLabelsTemplate Validates that the target template annotations and labels are both valid and that they do not contain reserved keys.
+func validateTargetMetadata(targetMetadata *trustapi.TargetMetadata, fldPath *field.Path) field.ErrorList {
+	el := field.ErrorList{}
+
+	if targetMetadata == nil {
+		return el
+	}
```
```suggestion
	if targetMetadata == nil {
		return nil
	}
	
	el := field.ErrorList{}
```

Nit

erikgb commented on pkg/webhook/validation.go:
```diff
@@ -231,9 +233,49 @@ func (v *validator) validate(ctx context.Context, obj runtime.Object) (admission
 	}
 
+	if bundle.Spec.Target.ConfigMap != nil {
+		el = append(el, validateTargetMetadata(bundle.Spec.Target.ConfigMap.Metadata, path.Child("target", "configMapTemplate"))...)
```
```suggestion
		el = append(el, validateTargetMetadata(bundle.Spec.Target.ConfigMap.Metadata, path.Child("target", "configMap", "metadata"))...)
```

erikgb commented on pkg/webhook/validation.go:
```diff
@@ -231,9 +233,49 @@ func (v *validator) validate(ctx context.Context, obj runtime.Object) (admission
+		el = append(el, validateTargetMetadata(bundle.Spec.Target.ConfigMap.Metadata, path.Child("target", "configMapTemplate"))...)
+	}
+	if bundle.Spec.Target.Secret != nil {
+		el = append(el, validateTargetMetadata(bundle.Spec.Target.Secret.Metadata, path.Child("target", "secretTemplate"))...)
```
```suggestion
		el = append(el, validateTargetMetadata(bundle.Spec.Target.Secret.Metadata, path.Child("target", "secret", "metadata"))...)
```

erikgb commented on test/integration/bundle/validation_test.go:
```diff
@@ -160,24 +160,54 @@ var _ = Describe("Bundle Validation", func() {
 	Context("Target", func() {
 		var (
-			selectorAccessor func(*trustapi.KeySelector)
+			selectorAccessor func(template *trustapi.TargetTemplate)
```
```suggestion
			selectorAccessor func(*trustapi.TargetTemplate)
```

Nit

erikgb reviewed (commented):
Excellent work, @terricain! 🚀  This PR is pretty close to being mergeable. I have added a few comments/suggestions, but mostly nits.

terricain commented:
Thanks again for the review 🙂, I've fixed the nits, also made the TrustTemplate getters nil-safe which makes things look cleaner.

Whats the procedure for the diff conversations, does the author or reviewer resolve them?

erikgb commented:
> Thanks again for the review 🙂, I've fixed the nits, also made the TrustTemplate getters nil-safe which makes things look cleaner.

Nice, thanks!

> Whats the procedure for the diff conversations, does the author or reviewer resolve them?

Not sure if we have a practice around this. Either works.

erikgb commented on pkg/apis/trust/v1alpha1/types_bundle.go:
```diff
@@ -212,13 +212,52 @@ type SourceObjectKeySelector struct {
+
+// GetAnnotations returns the annotations to be copied to the target or an empty map if there are no annotations.
+func (t *TargetTemplate) GetAnnotations() map[string]string {
+	if t == nil || t.Metadata == nil || t.Metadata.Annotations == nil {
```
```suggestion
	if t == nil || t.Metadata == nil {
```

Another nit, sorry for not spotting this last time.

erikgb commented on pkg/apis/trust/v1alpha1/types_bundle.go:
```diff
@@ -212,13 +212,52 @@ type SourceObjectKeySelector struct {
+
+// GetLabels returns the labels to be copied to the target or an empty map if there are no labels.
+func (t *TargetTemplate) GetLabels() map[string]string {
+	if t == nil || t.Metadata == nil || t.Metadata.Labels == nil {
```
```suggestion
	if t == nil || t.Metadata == nil {
```

Another nit, sorry for not spotting this last time.

erikgb commented on pkg/apis/trustmanager/v1alpha2/types_cluster_bundle.go:
```diff
@@ -212,13 +212,52 @@ type SourceObjectKeySelector struct {
+
+// GetAnnotations returns the annotations to be copied to the target or an empty map if there are no annotations.
+func (t *TargetTemplate) GetAnnotations() map[string]string {
+	if t == nil || t.Metadata == nil || t.Metadata.Annotations == nil {
```
```suggestion
	if t == nil || t.Metadata == nil {
```

Another nit, sorry for not spotting this last time.

erikgb commented on pkg/apis/trustmanager/v1alpha2/types_cluster_bundle.go:
```diff
@@ -212,13 +212,52 @@ type SourceObjectKeySelector struct {
+
+// GetLabels returns the labels to be copied to the target or an empty map if there are no labels.
+func (t *TargetTemplate) GetLabels() map[string]string {
+	if t == nil || t.Metadata == nil || t.Metadata.Labels == nil {
```
```suggestion
	if t == nil || t.Metadata == nil {
```

Another nit, sorry for not spotting this last time.

erikgb commented on pkg/bundle/internal/target/target.go:
```diff
@@ -442,6 +446,14 @@ func TrustBundleHash(data []byte, additionalFormats *trustapi.AdditionalFormats)
 	}
 
+	// Add Target annotations and labels to the hash so it becomes aware of changes and triggers an update.
+	for targetAnnotationKey, targetAnnotationValue := range target.GetAnnotations() {
+		_, _ = hash.Write([]byte(targetAnnotationKey + targetAnnotationValue))
+	}
+	for targetLabelKey, targetLabelValue := range target.GetLabels() {
+		_, _ = hash.Write([]byte(targetLabelKey + targetLabelValue))
+	}
```
```suggestion
	for k, v := range target.GetAnnotations() {
		_, _ = hash.Write([]byte(k + v))
	}
	for k, v := range target.GetLabels() {
		_, _ = hash.Write([]byte(k + v))
	}
```

More nitting, but it makes it a bit more compact, as the scope is so small.

erikgb reviewed (commented):
Thanks for the stamina on this. A couple of tiny nits I missed the last time.

Could you also please rebase and squash commits to make the commit history a bit cleaner?

/lgtm
/approve
/hold

erikgb commented:
> Have squashed the commits down to 2, mainly as one of the commits was authored by you, let me know if you want that squashed too.

Please squash everything into a single commit! I don't need credits for this work. You did most of the job!

terricain commented:
Ah you helped a fair bit 😄. Done.

erikgb reviewed (commented):
Great work! Thanks a lot for adding this new feature! 🚀 
Hoping for more nice contributions from you, @terricain!

/lgtm
/unhold

erikgb commented:
Do you minc creating a PR to update the website docs, @terricain? 🙏
